drv/transceivers-server/src/udp.rs

```diff
@@ -217,397 +215,572 @@ impl ServerImpl {
     }
 
     fn select_page(
         &mut self,
         page: Page,
-        mask: FpgaPortMasks,
-    ) -> Result<(), HwError> {
+        mask: LogicalPortMask,
+    ) -> ModuleResult {
         // Common to both CMIS and SFF-8636
         const BANK_SELECT: u8 = 0x7E;
         const PAGE_SELECT: u8 = 0x7F;
 
+        let mut result = ModuleResult::default();
+
         // We can always write the lower page; upper pages require modifying
         // registers in the transceiver to select it.
         if let Some(page) = page.page() {
-            self.transceivers
-                .set_i2c_write_buffer(&[page])
-                .map_err(|_e| HwError::PageSelectWriteBufFailed)?;
-            self.transceivers
-                .setup_i2c_write(PAGE_SELECT, 1, mask)
-                .map_err(|_e| HwError::PageSelectWriteFailed)?;
-            self.wait_and_check_i2c(mask)?;
+            result =
+                result.chain(self.transceivers.set_i2c_write_buffer(&[page]));
+            result = result.chain(self.transceivers.setup_i2c_write(
+                PAGE_SELECT,
+                1,
+                mask,
+            ));
+            result = result.chain(self.wait_and_check_i2c(result.success()));
+        } else {
+            // If the request is to the lower page it is always successful
+            result =
+                ModuleResult::new(mask, LogicalPortMask(0), LogicalPortMask(0))
+                    .unwrap();
         }
 
         if let Some(bank) = page.bank() {
-            self.transceivers
-                .set_i2c_write_buffer(&[bank])
-                .map_err(|_e| HwError::BankSelectWriteBufFailed)?;
-            self.transceivers
-                .setup_i2c_write(BANK_SELECT, 1, mask)
-                .map_err(|_e| HwError::BankSelectWriteFailed)?;
-            self.wait_and_check_i2c(mask)?;
+            result =
+                result.chain(self.transceivers.set_i2c_write_buffer(&[bank]));
+            result = result.chain(self.transceivers.setup_i2c_write(
+                BANK_SELECT,
+                1,
+                result.success(),
+            ));
+            result = result.chain(self.wait_and_check_i2c(result.success()));
         }
-        Ok(())
+        result
     }
 
-    fn wait_and_check_i2c(
-        &mut self,
-        mask: FpgaPortMasks,
-    ) -> Result<(), HwError> {
-        let err_mask = self
-            .transceivers
-            .wait_and_check_i2c(mask)
-            .map_err(|_e| HwError::WaitFailed)?;
-        if !err_mask.left.is_empty() {
-            // FPGA reported an I2C error
-            return Err(HwError::I2cError(PortMask(err_mask.left.get())));
-        } else if !err_mask.right.is_empty() {
-            return Err(HwError::I2cError(PortMask(err_mask.right.get())));
-        }
-        Ok(())
+    // Empty Ok() result if there were no I2C errors. The `LogicalPortMask`
+    // returned by the Err represents which ports had I2C errors.
```

Comment:
I think this comment is out of date now? It refers to a `Result`, but looks like we're now returning a `ModuleResult`.